File: src/modifiers/safeAreaInset.ts
import type { ReactNode } from 'react';
import { createModifier } from './createModifier';

export type SafeAreaInsetEdge = 'top' | 'bottom' | 'leading' | 'trailing';

export interface SafeAreaInsetParams {
  /// Edge to inset and place `content` against. `'top'` / `'bottom'` use a
  /// horizontal alignment; `'leading'` / `'trailing'` use a vertical one.
  edge: SafeAreaInsetEdge;
  /// Alignment of `content` along the edge. Defaults to `'center'`.
  alignment?: string;
  /// Spacing between `content` and the main view's content. Defaults to
  /// SwiftUI's automatic spacing when omitted.
  spacing?: number;
  /// View rendered in the reserved safe-area inset region.
  content: ReactNode;
}

/// SwiftUI `.safeAreaInset(edge:alignment:spacing:content:)`. Reserves space
/// along an edge of the view and places `content` there, shrinking the main
/// content's safe area to make room.
export function safeAreaInset(params: SafeAreaInsetParams) {
  return createModifier('safeAreaInset', params);
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/SafeAreaInset.swift
import SwiftUI

/// SwiftUI `.safeAreaInset(edge:alignment:spacing:content:)`. Reserves space
/// along an edge and places the slot `content` there, shrinking the main
/// content's safe area to fit. `top`/`bottom` use a `VerticalEdge` overload
/// (horizontal alignment); `leading`/`trailing` use a `HorizontalEdge`
/// overload (vertical alignment). Returns the view unchanged when no content
/// slot is provided.
enum RNWSafeAreaInsetModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("safeAreaInset") { view, params, ctx in
            guard let body = ctx.content(params.string("content")) else {
                return view
            }
            let spacing = params.cgFloat("spacing")
            switch params.string("edge") {
            case "bottom":
                return AnyView(view.safeAreaInset(
                    edge: .bottom,
                    alignment: RNWAlignmentParser.horizontal(
                        params.string("alignment")),
                    spacing: spacing
                ) { body })
            case "leading":
                return AnyView(view.safeAreaInset(
                    edge: .leading,
                    alignment: RNWAlignmentParser.vertical(
                        params.string("alignment")),
                    spacing: spacing
                ) { body })
            case "trailing":
                return AnyView(view.safeAreaInset(
                    edge: .trailing,
                    alignment: RNWAlignmentParser.vertical(
                        params.string("alignment")),
                    spacing: spacing
                ) { body })
            default: // "top"
                return AnyView(view.safeAreaInset(
                    edge: .top,
                    alignment: RNWAlignmentParser.horizontal(
                        params.string("alignment")),
                    spacing: spacing
                ) { body })
            }
        }
    }
}
